Try to cleanup the handling of cheribsd kernel configuration files

qwattash opened the pull request:
This attempts to have a separate class to fetch (and possibly generate) kernel configuration variants to build and run in cheribuild targets.

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -67,6 +71,277 @@ def _clear_dangerous_make_env_vars():
+    - Default configurations must select an unique (platform, kernel, flags) set, non-default
+      configurations may select multiple kernels.
+    """
+    # XXX-AM: For now hardcode the lists, ideally these may be generated
```
This is really nasty, this shouldn't be too hard to generate

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -67,6 +71,277 @@ def _clear_dangerous_make_env_vars():
+    def __repr__(self):
+        flags = []
+        for key,val in self.__dict__.items():
+            if key == "kernconf" or key == "platform" or key == "kABI":
+                continue
+            if val:
+                flags.append(key)
```
Surely there's a better way to do this kind of thing?

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -67,6 +71,277 @@ def _clear_dangerous_make_env_vars():
+
+
+class ConfigPlatform(IntEnum):
+    QEMU = 1 << 0
+    FVP = 1 << 1
+    GFE = 1 << 2
+    BERI = 1 << 3
+
+    def __str__(self):
+        platforms = []
+        for flag in ConfigPlatform:
+            if self & flag:
+                platforms.append(flag.name)
+        return "|".join(platforms)
+
+    @classmethod
+    def fpga_platforms(cls):
+        return (cls.BERI | cls.GFE)
```
Just make it a normal enum and use sets

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -67,6 +71,277 @@ def _clear_dangerous_make_env_vars():
 
+# XXX-AM: Replace this with CompilationTarget?
+class KernelABI(Enum):
+    NATIVE = "native"
```
What does native even mean for CHERI when hybrid and purecap exist too?

qwattash replied:
Probably `NOCHERI` is better here, I wanted to convey the fact that it is the target without CHERI extensions.

jrtc27 commented on pycheribuild/config/compilation_targets.py:
```diff
@@ -37,6 +37,7 @@
 from .loader import ConfigOptionBase
 from .target_info import (AutoVarInit, BasicCompilationTargets, CPUArchitecture, CrossCompileTarget, MipsFloatAbi,
                           TargetInfo)
+
```
```suggestion
```

jrtc27 replied:
still there

jrtc27 replied:
This *still* hasn't been deleted

qwattash replied:
I appear to have been very good at missing this :(

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -1259,45 +1534,65 @@ def __init__(self, config: CheriConfig):
+        configs = []
+        if self.build_purecap_kernels and kernABI != KernelABI.PURECAP:
+            # The default config has already been added if the default is purecap
+            # XXX-AM: should be build the hybrid kernel when the default is purecap?
```
Yes

jrtc27 commented on pycheribuild/projects/run_qemu.py:
```diff
@@ -413,28 +419,59 @@ def setup_config_options(cls, **kwargs):
     def process(self):
+        if self.list_kernels:
+            self._list_kernel_configs()
+            sys.exit(1)
```
1 is wrong, it ran successfully. But this is very inflexible; what happens if I want, say, `--run/list-kernels run-mips64-purecap run-riscv64-purecap`? IMO this should be a `return` not a `sys.exit`.

qwattash commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -1301,6 +1280,89 @@ def __init__(self, config: CheriConfig):
+        if self.compiling_for_mips(include_purecap=True):
+            purecap_kernels.append("CHERI_PURECAP_MALTA64")
+        elif self.compiling_for_riscv(include_purecap=True):
+            purecap_kernels.append("CHERI-PURECAP-QEMU-NODEBUG")
```
I think I copied whatever was being done in the original function, but I do agree that we probably don't want the benchmark kernel there.

qwattash replied:
I think with the current way of generating configuration entries this is fixed.

arichardson commented on pycheribuild/projects/run_qemu.py:
```diff
@@ -413,28 +418,71 @@ def setup_config_options(cls, **kwargs):
             self.disk_image = disk_image_class.get_instance(self).disk_image_path
 
+    def _valid_kernel_configs(self):
+        return self.source_project.get_kernel_configs(platform=ConfigPlatform.QEMU)
```
Should we just list all files `sys/<arch>/conf/` that don't contain a "." to get the full list?

qwattash replied:
There are configs in `sys/<arch>/conf` that we do not use or support in cheribuild and we never build. This would be sensible if we generated the set of kernel configurations by reading files in `sys/<arch>/conf` in the first place, which is possible, but I just went the other way around in the generator to avoid having a blacklist of files. I can try to have the generator read config files in `sys/<arch>` but would leave this call as is, so that it does not care about how we get configuration names

arichardson replied:
Sorry, I should have been a bit clearer in my comment: I meant printing the files in `sys/<arch>/conf` only for the --list-kernels option (possibly indicating which ones are automatically used by cheribuild), so that you can see the full list of strings that you can pass as the --kernel-config override. Does that make more sense?

qwattash replied:
Ah yes that can be done, however I think that in the cheribsd target it should list the kernels that can be given as default kernel overriede in --cheribsd/kernel-config, instead in qemu it should list the kernels that have been built and can be run (yet with the --qemu/kernel-config override, possibly changing the name of the qemu option to avoid confusion)

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -68,6 +71,333 @@ def _clear_dangerous_make_env_vars():
+    Cheribuild configuration descriptor for a CheriBSD kernel configuration file
+    """
+
+    def __init__(self, kernconf: str, platform: ConfigPlatform, **kwargs):
+        self.kernconf = kernconf
+        self.platform = {platform} if type(platform) != set else platform
+        self.kABI = kwargs.get("kABI", KernelABI.NOCHERI)
+        self.default = kwargs.get("default", False)
+        self.caprevoke = kwargs.get("caprevoke", False)
+        self.mfsroot = kwargs.get("mfsroot", False)
+        self.debug = kwargs.get("debug", False)
+        self.benchmark = kwargs.get("benchmark", False)
+        self.fuzzing = kwargs.get("fuzzing", False)
+        self.fett = kwargs.get("fett", False)
```
```suggestion
    def __init__(self, kernconf: str, platform: ConfigPlatform, kABI=KernelABI.NOCHERI, default=False, ...):
        self.kernconf = kernconf
        self.platform = {platform} if type(platform) != set else platform
        self.kABI = kABI
        self.default = default
        ...
```
Is there a reason we can't use default arguments here?

qwattash replied:
Not really, was easier to add new flags on the fly but perhaps less readable.

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -68,6 +71,333 @@ def _clear_dangerous_make_env_vars():
+            return "BERI"
+        return super().get_kabi_name(platform, kABI)
+
+    def get_flag_names(self, **kwargs):
```
default arguments?

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -68,6 +71,333 @@ def _clear_dangerous_make_env_vars():
+        self.fett_gfe_kernconf_components = OrderedDict(self.kernconf_components)
+        del self.fett_gfe_kernconf_components["platform_name"]
+
+    def get_flag_names(self, **kwargs):
```
Same here?

qwattash commented on pycheribuild/projects/run_qemu.py:
```diff
@@ -99,6 +99,8 @@ def setup_config_options(cls, default_ssh_port: int = None, **kwargs):
                                                                  "port. You can then use `ssh root@localhost -p $PORT` "
                                                                  "to connect to the VM")
+        cls.snapshot = cls.add_bool_option("snapshot", show_help=True,
+                                           help="Run in snapshot mode, changes to the disk image are non-persistent")
```
Makes sense, I used the same name as the qemu flag without much thought, but as Jess noted it is already confusing on the qemu side as there is also a separate snapshot functionality.

qwattash commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -1476,56 +1776,39 @@ def _build_and_install_kernel_binaries(self, build_cheribsd: BuildCHERIBSD, kern
+                                                  mfsroot=True)
+        configs += CheriBSDConfigTable.get_configs(xtarget, platforms, default_kABI,
+                                                   mfsroot=True, benchmark=True)
+        if self.build_purecap_kernels:
```
This is because when the default kernel is already purecap, the build-purecap-kernels flag adds the hybrid kernels to the build instead. This was changed originally after commenting with @jrtc27

arichardson replied:
Ah I missed that discussion. Maybe we should rename the field to something like "build_alternative_abi_kernels"?

arichardson commented on pycheribuild/config/compilation_targets.py:
```diff
@@ -517,23 +539,21 @@ def run_fpga_benchmark(self, benchmarks_dir: Path, *, output_file: str = None, b
-            basic_args.append(
-                "--kernel-img=" + str(mfs_kernel.installed_kernel_for_config(self.project, kernel_config)))
+            kernel_config = mfs_kernel.default_kernel_config(ConfigPlatform.BERI,
+                                                             benchmark=(not self.config.benchmark_with_debug_kernel))
```
```suggestion
                                                             benchmark=not self.config.benchmark_with_debug_kernel)
```

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -679,6 +995,18 @@ def clean(self) -> ThreadJoiner:
+        conf_dir = self.source_dir / "sys" / self.target_info.freebsd_target_arch_dir / "conf"
+        configs = conf_dir.glob("*")
+        blacklist = ["NOTES", "LINT", "DEFAULTS"]
+        self.info("Valid kernel configuration files for --freebsd/kernel-config:")
```
```suggestion
        self.info("Valid kernel configuration files for --" +self.target + "/kernel-config:")
```

arichardson replied:
Or even better since it ensures the second part won't become stale:
```suggestion
        self.info("Valid kernel configuration files for", self.get_config_option_name("kernel_config"))
```

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -1239,15 +1593,18 @@ def setup_config_options(cls, install_directory_help=None, **kwargs):
                                                  help="Path to an MFS root image to be embedded in the kernel for "
                                                       "booting")
 
+        cls.default_kernel = cls.add_config_option("default-kernel", show_help=True, _allow_unknown_targets=True,
```
Maybe this should be "default-kernel-abi" instead? Might make the --help confusing otherwise.

arichardson replied:
And I guess it also wants `only_add_for_targets`?

arichardson reviewed (approved):
A few minor points, but I think this is ready to land otherwise.

jrtc27 commented on pycheribuild/config/compilation_targets.py:
```diff
@@ -276,6 +277,22 @@ def triple_for_target(cls, target: "CrossCompileTarget", config: "CheriConfig",
         return target.cpu_architecture.value + common_suffix
 
+    @property
+    def freebsd_target_arch_dir(self):
+        """
+        Return the name of the target architecture machine-dependent code directory
+        in the source tree. (e.g. <arch> in sys/<arch>)
+        """
```
Isn't this TARGET_CPUTYPE, i.e. we should call this function freebsd_target_cputype?

jrtc27 replied:
Uh, TARGET_CPUARCH even (TARGET_CPUTYPE is like -mcpu).

jrtc27 replied:
This is what cheribsd.py's arch_build_flags does, too, so if you fix this to be `freebsd_target` then you can remove most of that `if` and leave behind just the CHERI handling.

qwattash commented:
Now should replace  #196 via the `--build-bench-kernels` option available for both cheribsd and MFS root targets.

arichardson commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -1630,50 +1636,63 @@ def get_default_kernel_abi(self):
+        if self.caprevoke_kernel:
+            combinations.append("caprevoke")
+        if self.build_fpga_kernels:
+            combinations.append("mfsroot")
```
This does not seem right. Should it be like this below?
```python3
        if self.build_fpga_kernels:
            configs += self._get_config_variants([ConfigPlatform.fpga_platforms()], kernABIs, combinations + ["mfsroot"])
```

qwattash replied:
I think you are right

arichardson reviewed (approved):
Still looks fine. Would be great to have a unit test that checks the kernel config list for a few configurations, but not essential right now.

qwattash commented:
> Still looks fine. Would be great to have a unit test that checks the kernel config list for a few configurations, but not essential right now.

Added some basic tests and changed the one in master to drop the `_BENCHMARK` variant unless `--build-bench-kernels` is given.

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -35,6 +36,9 @@
 import typing
 from pathlib import Path
 
+from collections import OrderedDict
+from enum import Enum
```
Why are these in a separate block?

qwattash replied:
No reason, stray newline.

jrtc27 commented on pycheribuild/projects/cross/cheribsd.py:
```diff
@@ -68,6 +72,345 @@ def _clear_dangerous_make_env_vars():
+    def make_config(self, platform, kABI, base_context=None, **kwargs):
+        kernconf_ctx = self._prepare_kernconf_context(platform, kABI, base_context=base_context, **kwargs)
+
+        valid_ctx_items = (v for v in kernconf_ctx.values() if v is not None and v != "")
```
If we return None for NOCHERI's kABI name can this lose the `!= ""`? Logically `""` would mean `-GFE` not `GFE`, for example.

qwattash replied:
Yes, fixed now.

arichardson commented on tests/test_argument_parsing.py:
```diff
@@ -617,9 +617,32 @@ def test_disk_image_path(target, expected_name):
+    pytest.param("cheribsd-mips64-purecap",
+                 ["--cheribsd/build-alternate-abi-kernels",
+                  "--cheribsd/default-kernel-abi", "purecap"],
+                 "CHERI_PURECAP_MALTA64", ["CHERI_MALTA64"]),
```
Could you add tests for morello too?

arichardson commented on tests/test_argument_parsing.py:
```diff
@@ -617,9 +617,32 @@ def test_disk_image_path(target, expected_name):
+    pytest.param("cheribsd-riscv64-purecap",
+                 ["--cheribsd/build-fett-kernels",
+                  "--cheribsd/build-alternate-abi-kernels"],
+                 "CHERI-QEMU-FETT", ["CHERI-QEMU", "CHERI-PURECAP-QEMU"]),
```
Maybe also add one RISC-V case to show that benchmark builds `CHERI-QEMU-NODEBUG`?

qwattash replied:
Turns out this was missing in the generator. Now fixed.

arichardson reviewed (approved):
LGTM once @jrtc27 is happy too.

jrtc27 commented:
The unnecessary whitespace diff lives on 20 days later...
